Publish openapi to gh pages

NicolasMassart opened the pull request:
## PR Description

Add CI step to generate Teku OpenAPI doc and publish it to Github pages.

Two steps are added:

- running Teku and extract the OpenAPI JSON spec file.
    Builds with the installDist task and runs Teku agains a testnet until the OpenAPI doc service is up and wget retrieves the JSON file. Wget retries 20 times by default so it lets some time for the service to be up.
- update the [github pages branch](https://github.com/PegaSysEng/teku/tree/gh-pages) used as source for the doc site.
    Uses a node.js container to run script available on the gh-pages branch in order to update the doc. It then commits and pushes the changes back to this branch using the SSH deploy key.

**IMPORTANT:** The tag part (creating a named version when a tag is created during release process) is not yet available.
Only the latest doc version will be built automatically from master for the moment until this tag CI part is solved.
If needed, a named version can be generated manually until we figure this out.
Until then, at least we will have the master automatically updating the latest doc version.
I will come back with a smaller PR to fix the tag part, but this large one will be merged and already updating the doc.

## Fixed Issue(s)

PegaSysEng/Misc#33

NicolasMassart committed:
test for CI OpenAPI publishing

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

NicolasMassart committed:
update pull test branch name

NicolasMassart committed:
remove redoc part as we will use it dynamically

NicolasMassart committed:
try with a medium executor as small seems to fail

NicolasMassart committed:
test

NicolasMassart committed:
test

NicolasMassart committed:
test

NicolasMassart committed:
test commit

NicolasMassart committed:
fix dir

NicolasMassart committed:
use another key

NicolasMassart committed:
tests for ssh

NicolasMassart committed:
typo

NicolasMassart committed:
test outside an executor

NicolasMassart committed:
remove branch filter test

NicolasMassart committed:
fix job in pipeline

NicolasMassart committed:
i'm not serious, need to focus!

NicolasMassart committed:
add executor back

benjamincburns committed:
fix ssh key, debug failing build step

Signed-off-by: Ben Burns <803016+benjamincburns@users.noreply.github.com>

NicolasMassart committed:
debugging

benjamincburns committed:
more fixes for CircleCI API publish config

Signed-off-by: Ben Burns <803016+benjamincburns@users.noreply.github.com>

benjamincburns committed:
gh-pages publish: only add the api.json file

NicolasMassart committed:
update executor and commit message

ajsutton commented on .circleci/config.yml:
```diff
@@ -266,57 +268,101 @@ jobs:
+          command: |
+            build/install/teku/bin/teku --rest-api-docs-enabled=true &
+            TEKU_PID=$!
+            wget -O api.json http://localhost:5051/swagger-docs
```
Isn't there a race condition here? We'd need to wait until the rest api service has actually started listening.

NicolasMassart replied:
wget does some retries.

NicolasMassart committed:
test split jobs

NicolasMassart committed:
fix ci

NicolasMassart committed:
update path

NicolasMassart committed:
ci process

NicolasMassart committed:
fix typo

NicolasMassart committed:
simple checkout

NicolasMassart committed:
uncomment ci

NicolasMassart committed:
update from master

and add TODO to change before merge

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

# Conflicts:
#	.circleci/config.yml

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

NicolasMassart committed:
use a testnet to be able to start Teku

bypass #1720

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

NicolasMassart committed:
update commit message to show it's from CI

NicolasMassart committed:
try tag filter

NicolasMassart committed:
Update commit message to include CI infos

NicolasMassart committed:
make sure doc is updated only if everything works

extract and publish API doc only if all tests and publishing worked
otherwise it could be updated even if some tests fails and teku is
not published

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

NicolasMassart committed:
make changes ready to merge on master

[skip ci] as we don't want the branch push to update the real doc
the update will be triggered when merged on master.
The tag part (creating a version when a tag is created) is not yet
available. Only the latest will be built automatically for the
moment until this tag CI part is solved. Until then at least we will
have the master updating the latest doc version. If needed a named
version can be generated manually until we figure this out.

NicolasMassart committed:
small change to update CI

I tried not to fire ci in the previous commit but CI status is needed
It will update the doc but as this branch is up to date with master
it's ok and even a good live test.
Doc rollback is easy in case of issue.

NicolasMassart committed:
fix typo

badly typed save shortcut ctrl+s added an "s"

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

ajsutton commented on .circleci/config.yml:
```diff
@@ -266,6 +272,76 @@ jobs:
+      - run:
+          name: Extract OpenAPI JSON
+          command: |
+            build/install/teku/bin/teku --network=schlesi --rest-api-docs-enabled=true &
+            TEKU_PID=$!
+            wget -O openapi.json http://localhost:5051/swagger-docs
+            kill $TEKU_PID
```
This has a race condition because Teku won't be listening the instant it's started.  We need something to retry for a period until the API is actually listening.  I would suggest splitting the script to do that into it's own file so that the circleci config stays as focussed on circleci specific stuff as possible.

NicolasMassart replied:
As [I already explained in a previous comment](https://github.com/PegaSysEng/teku/pull/1695#discussion_r419319799), there's no race condition here. Wget is retrying 20 times until it has a 200 code.

ajsutton replied:
Ah right, I didn't see that response because it got resolved before I had a chance to see it.

ajsutton commented on .circleci/config.yml:
```diff
@@ -266,6 +272,76 @@ jobs:
+            readonly BRANCH='gh-pages'
+            git checkout "$BRANCH" || git checkout -b "$BRANCH"
+            # pull might fail if the branch was just created
+            git pull || true
```
Would this be:
`git checkout "$BRANCH" && git pull || git checkout -b "$BRANCH"`

ie only pull the branch if it already existed.

NicolasMassart replied:
@benjamincburns you wrote these lines, do you have any comments on this please?

benjamincburns replied:
Ignore my last (deleted) comment - I had my wires crossed.

I think @ajsutton's suggestion is functionally equivalent to the code that we already have, but it'll save some noise in the build log if you change line 315 to what he has, and delete lines 316 and 317.

Also if the `git pull` fails in the one-liner that he suggested, it'll cause the `git checkout -b "$BRANCH"` statement to execute, which will also fail, causing the entire build to fail, however the build log might make the reason for the failure a little obscure. A better approach overall would be replace lines 315:317 with something like 

```bash
if [ git branch | grep -q "$BRANCH" ]; then
  git checkout "$BRANCH"
  git pull
else
  git checkout -b "$BRANCH"
fi
```

But this is likely overkill and I think any of the suggestions will work fine.

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

NicolasMassart committed:
update checkout per @benjamincburns comments

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages

NicolasMassart committed:
Merge branch 'master' into publish_openapi_to_gh-pages
